### DISCUSSION TITLE
Inserting large amount of data

### DISCUSSION THREAD

**Discussion by ghost** [supabase/supabase#511] [Questions] [answered]
I keep having my connection closed while trying to add a large amount of columns, is there any work around for this?

**Comment by steve-chavez** [comment, marked as answer]
**Edit**: `supabase-js/postgrest-js` now use the `columns` query parameter by default, no need to go to through the REST API for this.

@pearlslugs How are you adding the large amount of columns?

You could try inserting the big payload through the REST API. There's a special [columns query parameter](https://postgrest.org/en/v7.0.0/api.html#specifying-columns) that can help with large bulk inserts.

There are curl snippets in `https://app.supabase.io/project/<your_project_id>/api/default?resource=<your_table>`(click the API icon and then the Bash tab to go here).

**Reply by ghost** [reply]
Hey thank you for the response, I believe that the way I was doing it was actually the problem, I was attempting to do it from my front end in the fly just to create the database and that was a bad choice in my part, it would appear as though simply inserting it in a loop works fine if you do it properly.

**Reply by GodGardensGuns** [reply]
@steve-chavez - @pearlslugs teammate here. Do you have any resources for adding a content-type to the header of a supabase query? Also, I'm not seeing a Bash tab under the API tab in our database.

**Reply by steve-chavez** [reply]
@dayvista Hmm.. why do you want to change the `Content-Type` header? If you use `supabase-js` the requests payloads will be sent with a `Content-Type: application/json` already(check this [line](https://github.com/supabase/postgrest-js/blob/d56d69ab2dade6c98865530d9af18a1b54130ada/src/lib/types.ts#L74)).

> Also, I'm not seeing a Bash tab under the API tab in our database.

Can you double check? Here's how I see the Bash tab:

![2021-01-12-173502_1354x644_scrot](https://user-images.githubusercontent.com/1829294/104382766-b8920080-54fc-11eb-96c9-5962f46ec4b2.png)

**Reply by GodGardensGuns** [reply]
Seeing the bash tab now, thanks for pointing that out. The 'header' question is no longer relevant, as @pearlslugs ended up solving the problem by breaking the queries down into less data-intensive ones that wouldn't crash halfway through execution.

**Comment by thorwebdev** [comment]
Appending some more ideas that might be useful:

You might want to consider using the rpc function from the supabase-js client library to handle bulk processing as well (https://supabase.io/docs/reference/javascript/rpc#bulk-processing).

Outside of the client library, if you're looking to optimise the performance of bulk inserts on the database level, you can consider using a prepared INSERT statement and execute it as many times as you want (reduces the query planning time). You can also refer to this document on the Postgres site that explains how you can go around inserting large amounts of data into your db. (https://www.postgresql.org/docs/current/populate.html)

**Reply by steve-chavez** [reply]
As of now, [supabase-js' bulk insert](https://supabase.io/docs/reference/javascript/insert#bulk-create) should be pretty fast even for thousands of rows - this was also confirmed in https://github.com/supabase/supabase-js/issues/195#issuecomment-855317950.

I believe @burggraf has also dome some experiments on this, out-of-the-box bulk insert has been looking good enough for most use cases.

**Reply by burggraf** [reply]
We recently released an update to the `csv` / `tsv` import process in the Supabase dashboard and it uses the `Supabase Javascript Client` to import records to your PostgreSQL database.  I've personally tested it with millions of rows, and there's no technical limitation for the amount of data it will handle other than available disk space (as the process streams the `csv` file in chunks.) In my tests the speeds were very good, although I didn't really do any benchmark comparisons against any other methods.

**Reply by burggraf** [reply]
On thing to note --- I did dozens of speed tests for importing `csv` records for some huge tables, and I found, somewhat non-intuitively, that smaller batches worked better than larger batches.  It depends on your record length, of course, but I believe I found that a chunkSize of 0.25MB worked best for uploading records to the database.   We're using the `PapaParse` Javascript library to parse the `csv` and `tsv` files and in testing many chunkSize parameters, 0.25 megabytes seemed to be the sweet spot for fastest imports during my testing.
